habitat/utils/compare_profiles.py

```diff
@@ -0,0 +1,311 @@
+_collect_rollout_step                28,733         1,990          -469           -75
+wait_step                            19,079        19,079          -143          -143
+habitat_simulator.py step            18,523        18,523          -109          -109
+_worker_env send                     10,597        10,597          -414          -414
+nav.py update_metric                  9,145         9,145          -375          -375
+_update_agent                         7,559         1,357          -188            -9
+act (run policy)                      5,750         5,750          -106          -106
+evaluate_actions                      3,058         3,058           -95           -95
+batch_obs                             1,915         1,915          -144          -144
+before_step (for optimize)            1,681         1,681           -62           -62
+backward (for surrogate loss)         1,463         1,463           -22           -22
+"""
+import argparse
+import glob
+import os
+import sqlite3
+
+
+class Event:
+    """Python in-memory representation for an NVTX event from an sqlite
+    database"""
+
+    def __init__(self, name, thread_id, start, end):
+        self._name = name
+        self._thread_id = thread_id
+        self._start = start
+        self._end = end
+
+
+class SummaryItem:
+    """Summary item, for accumulating time for a set of events."""
+
+    def __init__(self):
+        self._time_exclusive = 0
+        self._time_inclusive = 0
+
+
+def get_sqlite_events(conn):
+    """Parse an sqlite database containing an NVTX_EVENTS table and return a
+    list of Events."""
+
+    events = []
+
+    # check if table exists
+    cursor = conn.execute(
+        "SELECT count(*) FROM sqlite_master WHERE type='table' AND name='NVTX_EVENTS'"
+    )
+    does_table_exist = cursor.fetchone()[0] > 0
+    if not does_table_exist:
+        return events
+
+    # select events
+    cursor = conn.execute(
+        "SELECT text, globalTid, start, end from NVTX_EVENTS"
+    )
+
+    for row in cursor:
+        events.append(Event(row[0], row[1], row[2], row[3]))
+
+    return events
+
+
+def create_summary_from_events(events):
+    """From a list of events, group by name and create summary items. Returns a
+    dictionary of items keyed by name."""
+    # sort by start time (ascending). For ties, sort by end time (descending). In this way,
+    # a (shorter) child event that starts at the same time as a (longer) parent event
+    # will occur later in the sort.
+    events.sort(reverse=True, key=lambda event: event._end)
+    events.sort(reverse=False, key=lambda event: event._start)
+
+    items = {}
+
+    for i in range(len(events)):
+
+        event = events[i]
+
+        item = items.setdefault(event._name, SummaryItem())
+
+        event_duration = event._end - event._start
+        item._time_inclusive += event_duration
+
+        exclusive_duration = 0
+
+        # iterate chronologically through later events. Our accumulated
```

Comment:
If you're trying to understand the logic here, it may be helpful to view the test example input at https://github.com/facebookresearch/habitat-lab/pull/454/files#diff-bad3b04d19297fe80d29e23c7e36696eR55 .